File: 2026-03/src/day_06/boj_17298/Main.cpp
#include <bits/stdc++.h>
using namespace std;

int arr[1000000];
int res[1000000];

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n;
    cin >> n;

    for (int i = 0; i < n; i++) cin >> arr[i];

    stack<int> st;
    for (int i = n - 1; i >= 0; i--) {
        int x = arr[i];

        while (!st.empty() && st.top() <= x) {
            st.pop();
        }

        res[i] = st.empty() ? -1 : st.top();
        st.push(x);
    }

    for (int i = 0; i < n; i++) {
        cout << res[i] << ' ';
    }
}

File: 2026-03/src/day_06/boj_17298/Main.java
package day_06.boj_17298;

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());

        int[] arr = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        Deque<Integer> stack = new ArrayDeque<>();
        int[] res = new int[n];

        for (int i = n - 1; i >= 0; i--) {
            int x = arr[i];

            while (!stack.isEmpty() && stack.peek() <= x) {
                stack.pop();
            }

            res[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(x);
        }

        for (int x : res) {
            sb.append(x).append(" ");
        }

        System.out.println(sb);
    }
}
